### ISSUE TITLE
Error with two column headers that have prefix in common

### ISSUE THREAD

**Issue by aghr** [nf-core/differentialabundance#249]
### Description of the bug

Pipeline works fine with samplesheet.csv with two columns with column headers `condition` and `Xcondition2`, but it throws error with `condition` and `condition2`. All other input files and parameters were identical.

Error shown below. 

### Command used and terminal output

```console
ERROR ~ Error executing process > 'NFCORE_DIFFERENTIALABUNDANCE:DIFFERENTIALABUNDANCE:CUSTOM_TABULARTO
GSEACLS (inv_vs_wt_e125)'                                                                             

Caused by:
  Process `NFCORE_DIFFERENTIALABUNDANCE:DIFFERENTIALABUNDANCE:CUSTOM_TABULARTOGSEACLS (inv_vs_wt_e125)
` terminated with an error exit status (2)

Command executed:

  cls_file=inv_vs_wt_e125.cls
  
  column_number=$(cat samplesheet.sample_metadata.tsv | head -n 1 | tr '\t' "\n" | grep -En "^conditio
n" | awk -F':' '{print $1}')
  classes=$(tail -n +2 samplesheet.sample_metadata.tsv | awk -F'\t' '{print $'$column_number'}')
  unique_classes=$(echo -e "$classes" | awk '!x[$0]++')
  
  echo -e "$(echo -e "$classes" | wc -l) $(echo -e "$unique_classes" | wc -l) 1" > $cls_file
  echo -e "#$(echo -e "$unique_classes" | tr '\n' ' ')" | sed "s/ $//" >> $cls_file
  echo -e "$classes" | tr '\n' ' ' | sed "s/ $//" >> $cls_file
  echo -e "\n" >> $cls_file
  
  
  cat <<-END_VERSIONS > versions.yml
  "NFCORE_DIFFERENTIALABUNDANCE:DIFFERENTIALABUNDANCE:CUSTOM_TABULARTOGSEACLS":
      bash: $(echo $(bash --version | grep -Eo 'version [[:alnum:].]+' | sed 's/version //'))
  END_VERSIONS

Command exit status:
  2

Command output:
  (empty)

Command error:
  awk: line 2: missing } near end of file

Work dir:
  XXX

Tip: you can replicate the issue by changing to the process work dir and entering the command `bash .command.run`

 -- Check '.nextflow.log' file for details
```

### System information

    version 23.10.0 build 5889 (created 15-10-2023 15:07 UTC (17:07 CEST))
    Linux desktop
    CentOS Linux release 7.7.1908
    local execution
    with singularity container
    version of nfcore diffabundance: 1.4.0

**Comment by BEFH** [comment]
I had this issue too. Probably can be fixed by changing this:

```bash
column_number=\$(cat $samples | head -n 1 | tr '$separator' "\\n" | grep -En "^$variable" | awk -F':' '{print \$1}')
```

to this:

```bash
column_number=\$(cat $samples | head -n 1 | tr '$separator' "\\n" | grep -En "^$variable\$" | awk -F':' '{print \$1}')
```

in https://github.com/nf-core/differentialabundance/blob/master/modules/nf-core/custom/tabulartogseacls/main.nf

But the file could probably use some fixing up in general. e.g. for that line:

```bash
column_number=\$(head -n 1 $samples | tr '$separator' "\\n" | grep -En "^$variable\$" | cut -d: -f1)
```

I'm also unsure how many backslashes it needs.

**Comment by BEFH** [comment]
Alternatively, you can replace line 30 and 31 with this:

```bash
classes=\$(awk -F '$separator' 'NR==1 { for (i=1; i<=NF; i++) if (\$i == $variable) {lnum = i; next}} 1 {print \$lnum}' $samples)
```

One-liner with no piping.

**Comment by asp8200** [comment]
Which branch was this error observed on? Is there a simple test nf-cmd to trigger the error?

**Comment by jenmuell** [comment]
To come back to the question of @asp8200, on which branch did you observe the error? I could only recreate the error on the main branch. If you pull the pipeline from the dev branch this should solve the issue. 
I could not recreate the error on the dev branch.

**Comment by BEFH** [comment]
Looks like it's already fixed here: https://github.com/nf-core/differentialabundance/blob/dev/modules%2Fnf-core%2Fcustom%2Ftabulartogseacls%2Fmain.nf#L30

But my suggested awk-only replacement of that line and the next might still be more robust. Should I bother with a pull-request or nah?

**Comment by jenmuell** [comment]
Hmm, I'm not familiar with the run time of awk. Could we run in some problems with the for-loop in your on-line awk option? Especially, with large datasets.

**Comment by WackerO** [comment]
@BEFH I will say I'm not an AWK expert, but if your code solution is more robust, do feel free to make a PR!
@jenmuell thanks for looking into this!

**Comment by grst** [comment]
Personally I think a saner solution would be to use a Python or R script - but I'll leave that up to whoever takes up this issue.

**Comment by nschcolnicov** [comment]
Hi @aghr @BEFH I'm trying to reproduce this error running the test dataset with the same input samplesheet but appending two additional columns "condition1" and "condition2", and I'm being unable to reproduce this error (more info about my run settings below).
Could you please share the command that you used, and any relevant input files? If you can't share the full input files, please share them with the values changed or just the first lines, so I can see the structure.

@grst Once I'm able to reproduce the error, I can put all of these commands into a python or R script, and move forward with updating the nf-core module.


RUN SETTINGS
This is what I'm running:
```
nextflow run nf-core/differentialabundance -r 1.4.0 -latest -profile docker,test -resume --input ../modified_test_files/SRP254919.samplesheet.csv --outdir results
```

The base samplesheet.csv file is this one `https://raw.githubusercontent.com/nf-core/test-datasets/modules/data/genomics/mus_musculus/rnaseq_expression/SRP254919.samplesheet.csv`, and I appended the two additional columns "condition1" and "condition2".

Here is my workdir for the CUSTOM_TABULARTOGSEACLS process:
[workdir (2).tar.gz](https://github.com/user-attachments/files/17738439/workdir.2.tar.gz)

**Comment by BEFH** [comment]
Try condition and condition1

**Comment by nschcolnicov** [comment]
@BEFH, Thank you for your response, it helped me found what I was missing. 
In order to reproduce this error, there must be a column in the samplesheet with the same prefix as the ["variable" variable ](https://github.com/nf-core/differentialabundance/blob/a3d664c12c4050bae2acc83b1c636dcc3546b9a5/modules/nf-core/custom/tabulartogseacls/main.nf#L25), which is defined as variable -> "args.variable" -> "meta.variable". The meta.variable comes from the contrasts file, which in the case of the test_profile, is labeled "treatment" for both rows. So by adding a column "treatment2" to the samplesheet I can now reproduce the error. 


@grst  Ill get started on the conversion now

**Comment by BEFH** [comment]
At least back when I first commented the issue was actually with a bad
regular expression in the script. I do not know if that is still the case,
but the regex wasn’t terminated with a $

**Comment by nschcolnicov** [comment]
I tested an implementation of this script using python, and this are the results for resource usage:

![Image](https://github.com/user-attachments/assets/301cfdf6-75f0-4767-9c34-90ab768c2ab6)


This is my POC of the module:
https://github.com/nf-core/differentialabundance/blob/issue_249/modules/nf-core/custom/tabulartogseacls/main.nf

@grst @jenmuell @WackerO @pinin4fjords  The script is definitely easier to understand but it has a pretty big decrease in resource efficiency. Do you think it is still worth to update this module into python?

If you agree with moving forward on the conversion, I'll update the modules nf-core repository

**Comment by pinin4fjords** [comment]
@nschcolnicov can we solve the problem without the rewrite?

**Comment by nschcolnicov** [comment]
> [@nschcolnicov](https://github.com/nschcolnicov) can we solve the problem without the rewrite?

Yes! The issue reported here is old, and has already been fixed in dev and 1.5.0; I can only reproduce it in 1.4.0.
The changes that we were discussing with @grst were related to simplifying the module so it can be more easily understood. 
It makes more sense to have this conversion as a separate ticket and close this one: https://github.com/nf-core/differentialabundance/issues/355
